docs/proposals/api-group.md

```diff
@@ -0,0 +1,150 @@
+
+## Server-side implementation
+
+1. resource handlers' URL:
+
+  We will force the URL to be in the form of prefix/group/version/...
+
+  Prefix is used to differentiate API paths from other paths like /healthz. All groups will use the same prefix="apis", except when backward compatibility requires otherwise. Specifically,
+
+    * for /api/v1, we set the prefix="api" (which is populated from cmd/kube-apiserver/app#APIServer.APIPrefix), group="", version="v1", so the URL remains to be /api/v1.
+
+    * for new kube API groups, we will set the prefix="apis" (we will add a field in type APIServer to hold this prefix), group=GROUP_NAME, version=VERSION. For example, the URL of the experimental resources will be /apis/experimental/v1alpha1.
+
+    * for OpenShift v1 API, because it's currently registered at /oapi/v1, to be backward compatible, OpenShift may set prefix="oapi", group="".
+
+    * for other new third-party API, they should also use the prefix="apis" and choose the group and version. This can be done through the thirdparty API plugin mechanism which is WIP in #13000.
+
+2. supporting API discovery:
+
+  * At /prefix (e.g., /apis), API server will return the supported groups and their versions in the form of "group/version".
+
+  * At /prefix/group, API server will return all supported versions of the group.
+
+  * At /prefix/group/version, API server will return all supported resources in this group, and whether each resource is namespaced.
+
+  We will design how to handle deeper path in other proposals.
+
+  * At /swaggerspec/prefix/group/version, API server will return the Swagger spec of that group/version.
+
+  * At /swaggerspec/prefix/major-version, API server will return a united Swagger spec of all groups in prefix/ at that version. This makes it easier to autogenerate clients based on Swagger spec.
+
+3. handling common API objects:
+
+  * top-level common API objects:
+
+    To handle the top-level API objects that are used by all groups, we either have to register them in all scheme.versionMap[<group/version>], or we can choose not to encode them to a version.
+
+    For Status (currently /pkg/api#Status), we plan to take the former approach, thus a client receiving the Status can extract the group/version information from the Status. Specifically, we plan to move /pkg/api#Status to a new package called `common`. The types.go and register.go files of every group/version has to define and register Status of that group/version. In most cases, API server will encode Status to the group/version embedded in the request URL, except when the group/version cannot be determined, then API server will send the Status with the field APIVersion="".
```

Comment:
@bgrant0607 OK, I'm convinced. The source of this was IRL chats between me and Chao. So the outcome is now that we'll fill in apiVersion only for legacy v1 clients.